### PR TITLE
L026: Handle DML statements and multiple levels of nesting

### LINKED ISSUE #2265: L026: Rule incorrectly flag column does not exist in `FROM` clause in an UPDATE statement.
## Expected Behaviour

L026 should not fail when a subquery in an UPDATE statement references a column from the UPDATE target.

## Observed Behaviour

L026 failed due to reference was not found in the FROM clause with the following error printed (When using `sample.sql` content below)

```
L:   7 | P:  28 | L026 | Reference 'my_table.id' refers to table/view not found
                       | in the FROM clause or found in parent subquery.
```

## Steps to Reproduce

1. Create `sample.sql` with the content below
```
UPDATE my_table
SET row_sum = (
    SELECT COUNT(*) AS row_sum
    FROM
        another_table
    WHERE
        another_table.id = my_table.id
);
```
2. Run SQLFluff by `sqlfluff lint sample.sql`

## Dialect

Default / Ansi (No dialect specified)

## Version
```
(.venv) ~/code/sqlfluff (main) $ sqlfluff --version
sqlfluff, version 0.9.0
```

```
(.venv) ~/code/sqlfluff (main) $ python --version
Python 3.9.9
```

## Configuration
Default. No customization.

### PR DESCRIPTION
**Pull request by barrywhart**
### Brief summary of the change made
Fixes #2265

Rewrite of L026:
* Supporting `UPDATE` and `DELETE` statements
* Support for arbitrarily nested subqueries (courtesy of `SelectCrawler`)
* Improve handling of multi-part table references (i.e. schema.table)

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: src/sqlfluff/rules/L025.py [hunk 1]
```diff
@@ -85,16 +85,18 @@ def _eval(self, context: RuleContext) -> EvalResultType:
     def _analyze_table_aliases(cls, query: L025Query, dialect: Dialect):
         # Get table aliases defined in query.
         for selectable in query.selectables:
-            select_info = get_select_statement_info(selectable.selectable, dialect)
```
Comment (barrywhart) [inline src/sqlfluff/rules/L025.py:88]: A few unrelated cleanups I made while adapting L025 for reuse as L026.

**Comment by jpy-git** [comment]
Good shout rewriting this, IMO I kind of hate all the rules that inherit from L020 as it makes it really hard to deal with the edge cases and results in having to shoehorn fixes in a non-ideal style. Much better having the separate logic 👍

File: src/sqlfluff/rules/L026.py [hunk 2]
```diff
@@ -100,4 +66,84 @@ def _eval(self, context: RuleContext) -> EvalResultType:
         ):
             return LintResult()
 
-        return super()._eval(context=context)
+        violations: List[LintResult] = []
+        start_types = ["select_statement", "delete_statement", "update_statement"]
```
Comment (jpy-git) [inline src/sqlfluff/rules/L026.py:70]: I had a look if I could trigger this false positive with the other DML statements too , INSERT and MERGE, but wasn't able to come up with a relevant case off the top of my head so seems fine to not include them here.

Have you also considered these and are happy to not include them in `start_types`?

(worst case they should be easy to add if a case arises)

File: src/sqlfluff/rules/L026.py [hunk 3]
```diff
@@ -100,4 +66,84 @@ def _eval(self, context: RuleContext) -> EvalResultType:
...
+            *start_types
+        ) and not context.functional.parent_stack.any(sp.is_type(*start_types)):
+            dml_target_table: Optional[str] = None
+            if not context.segment.is_type("select_statement"):
+                # Extract first table reference.
+                table_reference = next(
+                    context.segment.recursive_crawl("table_reference"), None
+                )
+                if table_reference:
+                    dml_target_table = table_reference.raw
```
Comment (jpy-git) [inline src/sqlfluff/rules/L026.py:81]: I think using raw here means the rule would get confused if the table references were more (or less) qualified in different places?

Bit of a toy example but this raises a lint:
```sql
use schema public;

delete from public.agent1
where exists(
    select customer.cust_id from customer
    where agent1.agent_code <> customer.agent_code);
```

It's a bit of a weird one as it seems niche and relies on the context that we're in the public schema already.
We can't just take the last identifier from the reference as we don't want to not flag:
```sql
delete from schema1.agent1
where exists(
    select customer.cust_id from customer
    where schema2.agent1.agent_code <> customer.agent_code);
```

Comment (jpy-git) [reply src/sqlfluff/rules/L026.py:81]: More relevant would be that something like this raises a lint:
```sql
delete from public.agent1
where exists(
    select customer.cust_id from customer
    where public.agent1.agent_code <> customer.agent_code);
```

File: src/sqlfluff/rules/L026.py (hunk 2 above)
Comment (barrywhart) [reply src/sqlfluff/rules/L026.py:70]: I agree, I think we are okay omitting them.

File: src/sqlfluff/rules/L026.py (hunk 3 above)
Comment (barrywhart) [reply src/sqlfluff/rules/L026.py:81]: You're right -- I'll take a look at this. Presumably, some existing rule(s) already handle similar cases.

**Comment by barrywhart** [comment]
@jpy-git: Thanks for the thorough review. Ready for another look, I think.

File: test/fixtures/rules/std_rule_cases/L026.yml [hunk 4]
```diff
@@ -110,3 +110,46 @@ test_pass_object_referenced_6:
...
+    where exists(
+        select customer.cust_id from customer
+        where agent1.agent_code <> customer.agent_code)
+
+test_pass_two_part_reference_9:
+  pass_str: |
+    delete from public.agent1
+    where exists(
+        select customer.cust_id from customer
+        where public.agent1.agent_code <> customer.agent_code)
```
Comment (jpy-git) [inline test/fixtures/rules/std_rule_cases/L026.yml:155]: So since the code is just looking at the table reference it doesn't flag this example:
```sql
delete from schema1.agent1
where exists(
    select customer.cust_id from customer
    where schema2.agent1.agent_code <> customer.agent_code)
```

`schema2.agent1` is not the same as `schema1.agent1`

I think we may need to check "do the table references match, if so do the schema references match, if so do the database references match?"
and in the event that one is more qualified than the other then allow it as long as the references present in both match to be lenient, like we have in `test_pass_two_part_reference_8` atm (i.e. `public.agent1` and `agent1` can match. Same for `some_db.some_schema.some_table` and `some_schema.some_table` and `some_table`)

Comment (barrywhart) [reply test/fixtures/rules/std_rule_cases/L026.yml:155]: I'll take another pass at it. This (i.e. robust name resolution, potentially ambiguous) seems like something we may (sooner or later) want to make reusable.

**Comment by jpy-git** [comment]
@barrywhart this isn't essential to be addressed within this PR but just flagging there's another open issue relating to L026: https://github.com/sqlfluff/sqlfluff/issues/1830

As I said doesn't need to fixed here, but worth bearing in mind whilst you update the rule 😄

File: src/sqlfluff/core/rules/reference.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
+"""Components for working with object and table references."""
+from typing import Sequence, Tuple
+
+
+def object_ref_matches_table(
```
Comment (barrywhart) [inline src/sqlfluff/core/rules/reference.py:5]: My thought process for adding this function was to write this algorithm in such a way that it was independent of other parts of SQLFluff (parser, `AliasInfo` class, etc.) This makes it easier to unit test, but importantly, I think it's also easier to understand in its pure form using **only** built-in Python types.

Does this seem like a good decision decision? Is this an appropriate home for this code?

File: src/sqlfluff/core/rules/reference.py [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+        return True
+    # Tricky case: If reference is shorter than target, it's allowed to match a
+    # suffix of the target. (Note this is an "optimistic" check, i.e. it assumes
+    # the ignored parts of the target don't matter. In a SQL context, this is
+    # basically assuming there was an earlier "USE <<database>>" or similar
+    # directive.
+    for pr in possible_references:
+        for t in targets:
+            if len(pr) < len(t) and pr == t[-len(pr) :]:
+                return True
```
Comment (jpy-git) [inline src/sqlfluff/core/rules/reference.py:24]: Shouldn't we also check the reverse situation, i.e. if target is shorter than the possible_reference and doesn't disagree?
```suggestion
            if (len(pr) < len(t) and pr == t[-len(pr) :]) or (
                len(t) < len(pr) and t == pr[-len(t) :]
            ):
                return True
```

File: src/sqlfluff/core/rules/reference.py (hunk 5 above)
Comment (jpy-git) [reply src/sqlfluff/core/rules/reference.py:5]: Yeah I'm happy with keeping this separate, makes our logic for determining ref matches self contained and well defined which is useful for rule writers 👍

File: test/fixtures/rules/std_rule_cases/L026.yml [hunk 7]
```diff
@@ -110,3 +110,59 @@ test_pass_object_referenced_6:
...
+    WHERE EXISTS(
+        SELECT customer.cust_id FROM customer
+        WHERE agent1.agent_code <> customer.agent_code);
+
+test_pass_two_part_reference_8:
+  pass_str: |
+    delete from public.agent1
+    where exists(
+        select customer.cust_id from customer
+        where agent1.agent_code <> customer.agent_code)
```
Comment (jpy-git) [inline test/fixtures/rules/std_rule_cases/L026.yml:148]: following on from my other review comment we should also try testing this the other way round: 
```yml
  pass_str: |
    delete from agent1
    where exists(
        select customer.cust_id from customer
        where public.agent1.agent_code <> customer.agent_code)
```

**Review by jpy-git — COMMENTED** [review]
Looks good, just need to verify the reverse check

File: src/sqlfluff/core/rules/reference.py (hunk 6 above)
Comment (barrywhart) [reply src/sqlfluff/core/rules/reference.py:24]: Will do!

File: test/fixtures/rules/std_rule_cases/L026.yml (hunk 7 above)
Comment (barrywhart) [reply test/fixtures/rules/std_rule_cases/L026.yml:148]: Incoming commit/push...

**Comment by barrywhart** [comment]
@jpy-git: Updated based on your last feedback. 👍

**Review by jpy-git — APPROVED** [review]
Looks great, nice work! LGTM 🚀 ✨
